File: Cpp/LeetCode128.LongestConsecutiveSequence.cpp
// Time Complexity: O(N)
// Space Complexity: O(N)

#include <unordered_set>
#include <vector>

class Solution {
public:
    int longestConsecutive(std::vector<int>& nums) {
        std::unordered_set<int> s;
        
        for (auto num:nums)
        {
            s.insert(num);
        }
        
        int ans = 0;
        
        for (auto num:nums)
        {
            if (!s.count(num-1)) // to ensure the consecutive sequence will not be processed again
            {
                int count = 1;
                int find = num+1;
                while (s.count(find))
                {
                    find++;
                    count++;
                }
                ans = std::max(ans, count);
            }
        }
        return ans;
    }
};

File: Python/LeetCode128.LongestConsecutiveSequence.py
class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:
        nums.sort()
        count = 1
        maxCount = 0 if len(nums) == 0 else 1
        for i in range(1, len(nums)):
            if nums[i] - nums[i-1] == 0:
                continue
            elif nums[i] - nums[i-1] == 1:
                count += 1
            else:
                count = 1
            maxCount = max(count, maxCount)
        return maxCount
